src/ensenso_driver.cpp

```diff
@@ -375,80 +465,218 @@ class EnsensoDriver
         pcl::toROSMsg(*cloud, cloud_msg);
         cloud_pub_.publish(cloud_msg);
       }
     }
 
-    void grabberCallback( const boost::shared_ptr<PairOfImages>& rawimages, const boost::shared_ptr<PairOfImages>& rectifiedimages)
+    void cloudRGBCallback( const boost::shared_ptr<PointCloudXYZRGBA>& cloud)
+    {
+      // Point cloud
+      if (cloud_pub_.getNumSubscribers() > 0)
+      {
+        ros::Time stamp;
+        //stamp is the same for all images/cloud
+        pcl_conversions::fromPCL(cloud->header.stamp, stamp);
+        publishTF(stamp);
+        cloud->header.frame_id = rgb_camera_frame_id_;
+        sensor_msgs::PointCloud2 cloud_msg;
+        cloud_msg.header.stamp = stamp;
+        pcl::toROSMsg(*cloud, cloud_msg);
+        cloud_pub_.publish(cloud_msg);
+      }
+    }
+
+    void imagesCallback( const boost::shared_ptr<PairOfImages>& rawimages, const boost::shared_ptr<PairOfImages>& rectifiedimages,
+                         const boost::shared_ptr<pcl::PCLGenImage<float> >& depthimage)
+    {
+      ros::Time stamp;
+      //stamp is the same for all images/cloud
+      pcl_conversions::fromPCL(depthimage->header.stamp, stamp);
+      // Get cameras info
+      sensor_msgs::CameraInfo linfo, rinfo, dinfo;
+      ensenso_ptr_->getCameraInfo("Left", linfo);
+      ensenso_ptr_->getCameraInfo("Right", rinfo);
+      ensenso_ptr_->getCameraInfo("Depth", dinfo);
+      linfo.header.stamp = stamp;
+      linfo.header.frame_id = camera_frame_id_;
+      rinfo.header.stamp = stamp;
+      rinfo.header.frame_id = camera_frame_id_;
+      dinfo.header.stamp = stamp;
+      dinfo.header.frame_id = camera_frame_id_;
+      // Images
+      if (l_raw_pub_.getNumSubscribers() > 0)
+        l_raw_pub_.publish(*toImageMsg(rawimages->first, stamp, camera_frame_id_), linfo, stamp);
+      if (r_raw_pub_.getNumSubscribers() > 0)
+        r_raw_pub_.publish(*toImageMsg(rawimages->second, stamp, camera_frame_id_), rinfo, stamp);
+      if (depth_pub_.getNumSubscribers() > 0)
+        depth_pub_.publish(*toImageMsg(*depthimage, stamp,camera_frame_id_), dinfo, stamp);
+      if (l_rectified_pub_.getNumSubscribers() > 0)
+        l_rectified_pub_.publish(toImageMsg(rectifiedimages->first, stamp, camera_frame_id_));
+      if (r_rectified_pub_.getNumSubscribers() > 0)
+        r_rectified_pub_.publish(toImageMsg(rectifiedimages->second, stamp, camera_frame_id_));
+      // Publish calibration pattern info (if any)
+      publishCalibrationPattern(stamp);
+    }
+
+    void imagesRGBCallback( const boost::shared_ptr<PairOfImages>& rawimages, const boost::shared_ptr<PairOfImages>& rectifiedimages,
+                            const boost::shared_ptr<PairOfImages>& rgbimages, const boost::shared_ptr<pcl::PCLGenImage<float> >& depthimage)
     {
-      ros::Time now = ros::Time::now();
+      ros::Time stamp;
+      //stamp is the same for all images/cloud
+      pcl_conversions::fromPCL(depthimage->header.stamp, stamp);
```

Comment:
This isn't working as expected when the `Cloud` streaming is disabled. 
Run the viewer without streaming the point cloud:
```
roslaunch ensenso viewer.launch serial:=YOUR_CAMERA_SERIAL
rosrun dynamic_reconfigure dynparam set /camera/ensenso_driver "{'Cloud':false, 'Projector':false, 'FrontLight':false, 'Images':true}"
```
Then, check these topics:
```
# Raw images
rostopic echo /camera/left/image_raw/header/stamp
rostopic echo /camera/right/image_raw/header/stamp
# Camera info
rostopic echo /camera/left/camera_info/header/stamp
rostopic echo /camera/right/camera_info/header/stamp
```
They all have stamp zero:
```
secs: 0
nsecs:         0
```

Is that the expected behavior?

Reply:
Since I take the depthimage from the cloud, the depthimage will not be calculated when cloud:=false.. I removed the depthimages from the imagesCallbacks. It will only be published now when Cloud AND Images parameters are set to true .